q2cli/core/usage.py

```diff
@@ -0,0 +1,231 @@
+
+        for opt_name, val in input_opts.items():
+            if opt_name in signature['inputs'].keys():
+                inputs[opt_name] = (val, signature['inputs'][opt_name])
+            elif opt_name in signature['params'].keys():
+                if is_iterable(val):
+                    val = list(val)
+                params[opt_name] = (val, signature['params'][opt_name])
+            elif opt_name in signature['mds'].keys():
+                mds[opt_name] = (val, signature['mds'][opt_name])
+
+        for opt_name, val in output_opts.items():
+            outputs[opt_name] = (val, signature['outputs'][opt_name])
+
+        return inputs, params, mds, outputs
+
+
+class CLIRenderer:
+    def __init__(self, records):
+        self.cache_records = records
+
+    def render(self):
+        if len(self.cache_records) == 0:
+            yield 'No examples have been registered for this action yet.'
+        else:
+            for record in self.cache_records:
+                yield self.dispatch(record)
+
+    def dispatch(self, record):
+        source = record['source']
+
+        if source == 'comment':
+            return self.template_comment(record['value'])
+        elif source == 'action':
+            return self.template_action(
+                record['value']['plugin_id'],
+                record['value']['action_id'],
+                record['value']['inputs'],
+                record['value']['params'],
+                record['value']['mds'],
+                record['value']['outputs'],
+            )
+        else:
+            raise Exception('uh oh')
+
+    def template_comment(self, comment):
+        return f'# {comment}'
+
+    def template_action(self, plugin_id, action_id,
+                        inputs, params, mds, outputs):
+        templates = [
+            *self._template_inputs(inputs),
+            *self._template_parameters(params),
+            *self._template_metadata(mds),
+            *self._template_outputs(outputs),
+        ]
+
+        base_cmd = to_cli_name(f'qiime {plugin_id} {action_id}')
+        action_t = self._format_templates(base_cmd, templates)
+        return action_t
+
+    def _format_templates(self, command, templates):
+        wrapper = textwrap.TextWrapper(initial_indent=' ' * 4)
+        templates = [command] + [wrapper.fill(t) for t in templates]
+        # TODO: double-check that string escaping is working
+        return ' \\\n'.join(templates)
+
+    def _template_inputs(self, input_opts):
+        inputs = []
+        for opt_name, (ref, _) in input_opts.items():
+            refs = ref if isinstance(ref, list) else [ref]
+            for ref in refs:
+                opt_name = to_cli_name(opt_name)
+                inputs.append(f'--i-{opt_name} {ref}.qza')
+        return inputs
+
+    def _template_parameters(self, param_opts):
+        params = []
+        for opt_name, (val, _) in param_opts.items():
+            # TODO: circle back on sets
```

Comment:
Native Python sets aren't JSON compatible, so prior to serializing to JSON for the CLI Cache, we convert any collections into `list` (https://github.com/qiime2/q2cli/pull/239/files#diff-0e5556f314d00b6f2414122af6ed2260R103-R104). The templater here will need to unpack that info. Strategies include looking at the `qiime_type` (its the `_` above), or, serializing the original collection type (`_destructure_opts`), or, somehow looking up the record (you would need to do that in `CLIUsage` though, the records will no longer exist here in `CLIRenderer`).